Publish Github releases from Travis

mynameisfiber opened the pull request:
This release introduces some config to automatically build deploy a `.deb` and `.dmg` on the github release page using travis. The build only happens when a commit is tagged and it uses the stable version of rust.

TODO:
- Mechanism for changelog in both debian release and github release
- Mechanism for manpages in debian release
- Potential PPA upload for debian release?

BEFORE MERGE:
- In `.travis.yml`, the github api key needs to be changed to one that has write permissions for this repo

AFTER MERGE:
- Set cron task in travis to create nightlies

@chrisduerr ^

This pull request is built off of #1324

chrisduerr commented:
Running a complete `cargo build --release` before actually building seems like a bad way to do this. It would probably make much more sense to just run `cargo install cargo-deb` after the build process.

mynameisfiber commented:
Sounds good to me! I'm not that familiar with the `cargo` build process... what is the difference between running the install for `cargo-deb` before/after building a release?

chrisduerr commented:
It's not really about the `cargo-deb` build process. It's just that running the actual build before the build process will mean that you don't get the real build in the actual test. Because the real build already happened and you just get the cached version of it. It should still display all errors if there are some but it's definitely not optimal.

chrisduerr commented:
I've actually looked at it again and it seems like I got confused for a second. You have the stuff in the `before_deploy` section, not before the build process. In this case it's completely fine the way it is right now (and the way it was before). Sorry for the confusion.

chrisduerr commented on .travis.yml:
```diff
@@ -23,7 +24,32 @@ matrix:
+  on:
+    rust: stable
+    all_branches: true
+    condition: '$TRAVIS_TAG != "" || $TRAVIS_EVENT_TYPE = "cron"'
```
Looking right at the top of the travis documentation (https://docs.travis-ci.com/user/deployment/releases/), it should be possible to just use `on: tags: true`.

mynameisfiber replied:
Using `tags: true` wouldn't work for building nightlies. I can definitely set it, but then any travis-based cron job wouldn't trigger the deploy process. I don't mind either way though.

chrisduerr replied:
What do you mean by `travis-based cron job`? We'd only want to release `.deb` files for stable releases that are manually tagged through `git`. I don't think providing nightly `.deb`s should be a focus for now, for nightly compiling yourself seems like the way to go.

If we would release nightlies instead of only commits tagged on github we'd probably run into issues with a ton of people running nightly instead of the stable version. It seems like a smart choice to try and get most people to use the tagged releases.

NickeZ replied:
Having a PPA with nightlies would be great though. It takes quite a while to rebuild Alacritty if you update the rust compiler and don't have the fastest machine in the world.

chrisduerr replied:
Yeah that's a good argument @NickeZ, but with releases on github I'd stick with stable releases first. I'm not quite sure how it works either, but I think with github the latest version (in that case nightly) would always be the first element and thus reduce the visibility of stable releases.

If there is a convinient way to do this, it would be fine though. However it should still be disabled on nightly builds, maybe if we want to push nightlies it should be a separate job that just gets automatically triggered every night to release them. But I'd say something like that is so far out that I wouldn't worry too much about it yet.

chrisduerr commented on .travis.yml:
```diff
@@ -23,7 +24,32 @@ matrix:
+  api_key:
+    secure: o2Q39A2JAKeMEYD0dFJTpYJMpAsHYcUa4DIXApqn/ygXQzcfPpnLelCWb70zOlICIk3ICETlzeLfF7s/q802jy5aEFLbbmIp+rT0SWlRiShuQqLPkwlK7Cqh2nUAVPoozYn1y9ha8kozOL3xgDxI9RGHMC+7ZV9duVLWUqz5Ah7qBoOx9QemDMrdPJ7zi8vFvJp4oW3EV7wQRRfEmFkqOAxbAWhLeTNGVR9ECjUZIgXWRz1kjqVz6dJrLauH/D068RSV/7JM1dlzrMkBsmRe8xVOTnAiHIoGMB4UaUTLz8hJNH+KUmXfprfBKuFEQ7jDnYq+6/y200VYuRfRsU2EOqYjzHXmWD6E3vkHhVdrUUfFmG9tfAzIXFZxs0LVIszWgqq ...
+  skip_cleanup: true
+  overwrite: true
```
I think it would be better to remove this line. Creating a build manually when the process fails and overwriting existing releases is no problem. However once a file is overwritten it can't be undone.

mynameisfiber replied:
:+1:

chrisduerr commented on .travis.yml:
```diff
@@ -23,7 +24,32 @@ matrix:
+  skip_cleanup: true
+  overwrite: true
+  file_glob: true
+  file: $DEPLOY_FILE
```
Since you're already working on this, it might make sense to also include uploading the binary directly. So this PR would just do everything we need to deploy binaries to github.

You can upload multiple files like this:
```
file:
    - "FILE 1"
    - "FILE 2"
```

Since you have a `cargo build --release` in the `before_deploy`, you could just upload that file directly. Regarding naming, I'd propose something like this: `alacritty-0.18.3` and `alacritty-0.18.3.deb`. It should be possible to do this using the tag name, so just `alacritty-TAG` for the binary and `alacritty-TAG.deb` for the `.deb` file.

mynameisfiber replied:
Uploading the binary is a good idea. I suggest we do `alacritty-TAG-ARCH`.

mynameisfiber commented:
No worries! Travis doesn't make this stuff clear so it's easy to get confused (as you can tell by the 36 attempted builds on my local fork of this repo)

mynameisfiber commented:
Okay... I addressed your above comments. You'll be able to see the file names [here](https://github.com/mynameisfiber/alacritty/releases/tag/v0.1.0-3) once the deploy is finished.

chrisduerr commented:
Those filenames look nice! This PR is achieving more than I initially expected. :D

Looks like we should be able to properly setup all github releases so we can just rely on that to work automagically once we start tagging, thanks!

thedrow commented:
LGTM but needs to be rebased.

chrisduerr commented:
I've just merged #1324, so there shouldn't be any other PR holding this back. However this will probably be merged after scrollback is released when we have proper versioning and all the other PRs can be cleaned up too. I'd like jwilm to check this out before merging.

Don't worry about the conflicts, I can resolve them myself before merging. Resolving these now will just lead to new conflicts being introduced in the future.

mynameisfiber commented:
Remember that the secret in the cargo file needs to correspond to an api
key that can create new releases!

On Sun, Jul 15, 2018, 9:03 AM Christian Duerr <notifications@github.com>
wrote:

> I've just merged #1324 <https://github.com/jwilm/alacritty/pull/1324>, so
> there shouldn't be any other PR holding this back. However this will
> probably be merged after scrollback is released when we have proper
> versioning and all the other PRs can be cleaned up too. I'd like jwilm to
> check this out before merging.
>
> Don't worry about the conflicts, I can resolve them myself before merging.
> Resolving these now will just lead to new conflicts being introduced in the
> future.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jwilm/alacritty/pull/1329#issuecomment-405090673>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAC5CqmmNuLWdQgbvMcnvbD4sD3W0LWaks5uGz2cgaJpZM4UNO2b>
> .
>

chrisduerr commented:
I'll be taking a stab at getting this back on track for all platforms myself.

Hopefully that should allow us to get binaries on all platforms working soon.

chrisduerr commented:
Note that everything should work now, except for the security key which still needs to be changed.

However there are still two other remaining issues:

Since Linux binaries aren't statically linked, it is not possible to execute the produced binary. This can be resolved by making use of musl for static linking. See `ripgrep` as an example for that:
https://github.com/BurntSushi/ripgrep/blob/master/.travis.yml

Currently the `.deb` version is called `Alacritty-v0.5.5-x86_64.deb`, however there were some issues with the name of the `.deb` file during installation. Looking at `ripgrep`, it also adds a different platform indicator for the `.deb` file compared to the other target systems. So it's probably best to copy that behavior to make sure everything works properly.

chrisduerr commented:
This PR should be good to go now. For visibility here's the last commit message, which should describe all changes that have been made:

```
A few changes have been made to the automatic building of Alacritty
binaries with Travis.

Since providing binaries on Linux isn't easily possible because of
dynamic linking, I've decided to leave this up to package maintainers.
Since the .deb format allows us to target a relatively large audience
with hopefully consistent library locations, this is the only Linux
binary which will be provided.

The main travis sections (install/script/before_deploy) have also been
moved out of the .travis.yml to make it easier to read, maintain and
extend the different steps of the CI process.

Since checking for the Rust version in CI is enough to know if clippy
should be used or not, the environment variable `CLIPPY` has also been
removed, which further allowed simplifying the CI process.

Besides the executables, some auxillary files are now also published as
part of a release when they have changed since the last tagged Alacritty
release. This should make it clear for returning users when a new
version of a specific auxillary file is required.
```

The only thing left now is for @jwilm to generate a Github access token with the `public_repo` permission, encrypt it using the command `travis encrypt -r jwilm/alacritty <GITHUB_TOKEN>` and then replace the current value in the `.travis.yml` file.

chrisduerr commented:
With the latest changes the CI process should now produce binaries which work on Debian, Fedora, Archlinux and Ubuntu without having to provide multiple versions.
